src/objective/rank_obj.cu

```diff
@@ -0,0 +1,651 @@
+      atomicAdd(out_pos_gpair + 1, in_pos_gpair.GetHess());
+
+      float *out_neg_gpair = reinterpret_cast<float *>(&out_gpair[original_pos[neg_idx]]);
+      const GradientPair in_neg_gpair(-g * weight, 2.0f * weight * h);
+      atomicAdd(out_neg_gpair, in_neg_gpair.GetGrad());
+      atomicAdd(out_neg_gpair + 1, in_neg_gpair.GetHess());
+    });
+  }
+};
+#endif
+
+// objective for lambda rank
+template <typename LambdaWeightComputerT>
+class LambdaRankObj : public ObjFunction {
+ public:
+  void Configure(const std::vector<std::pair<std::string, std::string> >& args) override {
+    param_.InitAllowUnknown(args);
+  }
+
+  void GetGradient(const HostDeviceVector<bst_float>& preds,
+                   const MetaInfo& info,
+                   int iter,
+                   HostDeviceVector<GradientPair>* out_gpair) override {
+    CHECK_EQ(preds.Size(), info.labels_.Size()) << "label size predict size not match";
+
+    // quick consistency when group is not available
+    std::vector<unsigned> tgptr(2, 0); tgptr[1] = static_cast<unsigned>(info.labels_.Size());
+    const std::vector<unsigned> &gptr = info.group_ptr_.size() == 0 ? tgptr : info.group_ptr_;
+    CHECK(gptr.size() != 0 && gptr.back() == info.labels_.Size())
+        << "group structure not consistent with #rows";
+
+    const auto ngroup = static_cast<bst_omp_uint>(gptr.size() - 1);
+    bst_float sum_weights = 0;
+    for (bst_omp_uint k = 0; k < ngroup; ++k) {
+      sum_weights += info.GetWeight(k);
+    }
+    bst_float weight_normalization_factor = ngroup / sum_weights;
+
+#if defined(__CUDACC__)
+    // For now, we only support pairwise ranking computation on GPU.
+    // Check if we have a GPU assignment; else, revert back to CPU
+    auto device = tparam_->gpu_id;
+    if (device >= 0 && LambdaWeightComputerT::SupportOnGPU()) {
+      LOG(DEBUG) << "Computing pairwise gradients on GPU.";
+      dh::safe_cuda(cudaSetDevice(device));
+
+      // Set the device ID and copy them to the device
+      out_gpair->SetDevice(device);
+      info.labels_.SetDevice(device);
+      preds.SetDevice(device);
+      info.weights_.SetDevice(device);
+
+      out_gpair->Resize(preds.Size());
+
+      auto d_preds = preds.ConstDevicePointer();
+      auto d_gpair = out_gpair->DevicePointer();
+
+      if (!slist_) {
+        slist_.reset(new SortedLabelList(device, param_));
+      }
+
+      // Create segments based on group info
+      slist_->InitWithTrainingInfo(gptr);
+
+      // Sort the labels within the groups on the device
+      slist_->Sort(info.labels_);
+
+      // Initialize the gradients next
+      out_gpair->Fill(GradientPair(0.0f, 0.0f));
+
+      // Finally, compute the gradients
+      slist_->ComputeGradients(d_preds, d_gpair,
+                               info.weights_, weight_normalization_factor);
+
+      // Wait until the computations done by the kernel is complete
+      dh::safe_cuda(cudaStreamSynchronize(nullptr));
```

Comment:
to block the host thread until the kernel has finished execution.

Reply:
Are you sure you need to do this? I think if someone uses the gradients on the host it will synchronise at that point. Unnecessary synchronisation can cost quite a bit of performance.

Reply:
this should only be called if the gradients are computed on the gpu and i did this to wait for the kernel to complete before returning from the `GetGradient` method. by not blocking the thread until the kernel completes execution, wouldn't we be bailing prematurely? subsequent kernels on the default stream should be blocked even if we do not do that; however, i did this for this method to be fully self contained, and that it returns only after everything is computed (i.e. the out params are fully populated).